lib/ansible/plugins/inventory/hcloud.py

```diff
@@ -0,0 +1,204 @@
+# Copyright (c) 2019 Hetzner Cloud GmbH <info@hetzner-cloud.de>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+
+__metaclass__ = type
+
+DOCUMENTATION = r"""
+    name: hcloud
+    plugin_type: inventory
+    authors:
+      - Lukas Kaemmerling (@lkaemmerling)
+    short_description: Ansible dynamic inventory plugin for the Hetzner Cloud.
+    version_added: "2.8"
+    requirements:
+        - python >= 2.7
+        - hcloud-python >= 1.0.0
+    description:
+        - Reads inventories from the Hetzner Cloud API.
+        - Uses a YAML configuration file that ends with hcloud.(yml|yaml).
+    options:
+        plugin:
+            description: marks this as an instance of the "hcloud" plugin
+            required: true
+            choices: ["hcloud"]
+        token:
+            description: The Hetzner Cloud API Token.
+            required: true
+            env:
+                - name: HCLOUD_TOKEN
+        connect_with:
+            description: Connect to the server using the value from this field.
+            default: public_ipv4
+            type: str
+            choices:
+                - public_ipv4
+                - hostname
+                - ipv4_dns_ptr
+        locations:
+          description: Populate inventory with instances in this location.
+          default: []
+          type: list
+          required: false
+        types:
+          description: Populate inventory with instances with this type.
+          default: []
+          type: list
+          required: false
+        images:
+          description: Populate inventory with instances with this image name, only available for system images.
+          default: []
+          type: list
+          required: false
+        label_selector:
+          description: Populate inventory with instances with this label.
+          default: ""
+          type: str
+          required: false
+"""
+
+EXAMPLES = r"""
+# Minimal example. `HCLOUD_TOKEN` is exposed in environment.
+plugin: hcloud
+
+# Example with locations, types, groups and token
+plugin: hcloud
+token: foobar
+locations:
+  - nbg1
+types:
+  - cx11
+"""
+
+import os
+from ansible.errors import AnsibleError, AnsibleParserError
+from ansible.module_utils._text import to_native
+from ansible.plugins.inventory import BaseInventoryPlugin
+from ansible.release import __version__
+
+try:
+    from hcloud import hcloud
+except ImportError:
+    raise AnsibleError("The Hetzner Cloud dynamic inventory plugin requires hcloud-python.")
+
+
+class InventoryModule(BaseInventoryPlugin):
+    NAME = "hcloud"
+
+    def _configure_hcloud_client(self):
+        self.api_token = self.get_option("token")
```

Comment:
How about
```suggestion
        self.api_token = self.get_option("token") or os.getenv("HCLOUD_TOKEN")
```
to reduce nesting?